### PR TITLE
feat: add entity catalogs, AddContentConfig types & helper functions

### PR DESCRIPTION
**Pull request by dbouwman**
1. Description:

- add `IWithCatalogs { catalogs?: IHubCatalog[] }` to Site, Project, Initiative, Event, Discussion
- mark `IWithCatalog` as deprecated 
- base property map will copy `data.catalogs` -> `.catalogs` if present
- add `searchEntityCatalogs(...)`
```js
searchEntityCatalogs(
  entity: HubEntity,
  query: string | IQuery,
  options: IPagingOptions & ISortOptions,
  context: IArcGISContext
): Promise<ICatalogSearchResponse[]>
```
- add `searchCatalogs(...)`
```js
export async function searchCatalogs(
  catalogs: IHubCatalog[],
  query: string | IQuery,
  options: IPagingOptions & ISortOptions, // only `num` is currently used, 
  context: IArcGISContext
): Promise<ICatalogSearchResponse[]>
```
- add `ICatalogSearchResponse`
```js
export interface ICatalogSearchResponse {
  catalogTitle: string;
  collectionResults: ISearchResponseHash;
  scopeResults?: ISearchResponseHash;
}
```
- add `getQueryContentConfig`
```
export function getQueryContentConfig(
  query: IQuery,
  context: IArcGISContext
): IQueryContentConfig
```

- add `getCatalogContentConfig`
```
export function getCatalogContentConfig(
  catalog: IHubCatalog,
  context: IArcGISContext
): ICatalogContentConfig
```

- add `getContentConfig` 
```
export function getAddContentConfig(
  context: IArcGISContext,
  catalogOrQuery?: IQuery | IHubCatalog
): IAddContentWorkflowConfig
```
- passing in just a context, returns the types the user can create (based on licensing + privs)
- passing in a query returns the types in the query the user can create (based on licensing + privs) and the groups they can share them to, as well as the "existing" types they can choose to add to the groups, along with a query that defines the choosable things
- passing a catalog returns the types the user can create (based on licensing + privs) and the groups that are in the catalog, which they can share do.
- This all lines up with changes to `arcgis-hub-add-content` component that leverages all this

- DONE: Add some "e2e" validation tests to ensure it actually works and to check performance
- DONE:  `searchCatalogs` will search "scopes without collections" 
- FUTURE: entity migrations to move any existing `.catalog` structures into `.catalogs` with a title of "Content", 
- FUTURE: property mapping for Event

1. Instructions for testing:
- run tests

1. Closes Issues: #10684, #10918

1. [x] Updated meaningful TSDoc to methods including Parameters and Returns, see [Documentation Guide](https://esri.github.io/hub-components/storybook/?path=/story/guides-documentation--page)

1. [x] used semantic commit messages
  
1. [x] PR title follows semantic commit format (**CRITICAL** if the title is not in a semantic format, the release automation will not run!)

1. [ ] updated `peerDependencies` as needed. **CRITICAL** our automated release system can **not** be counted on to update `peerDependencies` so we _must_ do it manually in our PRs when needed. See the [updating peerDependencies](/RELEASE.md#Updating-peerDependencies) section of the release instructions for more details.

### CODE REVIEW

File: packages/common/src/search/searchCatalogs.ts [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+        return {
+          catalogTitle: cat.title,
+          collectionResults: response,
+        } as ICatalogSearchResponse;
+      });
+    const promiseArray = [collectionPromise];
+    // for this catalog, find any scopes that do not have a collection
+    const collectionTargets = cat.collections
+      .map((collection) => collection.targetEntity)
+      .filter((v, i, self) => self.indexOf(v));
```
Comment (mjuniper) [inline packages/common/src/search/searchCatalogs.ts:54]: So this gives us an array of targetEntities with duplicates removed? It took me a while to understand what was going on with the filter part. Maybe a comment would help our future selves?

**Review by mjuniper — APPROVED** [review]

File: packages/common/src/search/searchCatalogs.ts (hunk 1 above)
Comment (dbouwman) [reply packages/common/src/search/searchCatalogs.ts:54]: I'll re-work so it's more obvious - this was some generated code, which while correct, is not as clear as...
```
const collectionTargets = mapBy("targetEntity", cat.collections)
     .filter(unique);
```

File: packages/common/src/search/getQueryContentConfig.ts [hunk 2]
```diff
@@ -0,0 +1,349 @@
+import { IArcGISContext } from "../ArcGISContext";
```
Comment (dbouwman) [inline packages/common/src/search/getQueryContentConfig.ts:1]: @mjuniper This is the meat of the "config" stuff. The catalog version of this fn, delegates here

File: packages/common/src/search/searchCatalogs.ts [hunk 3]
```diff
@@ -0,0 +1,94 @@
+import { IArcGISContext } from "../ArcGISContext";
```
Comment (dbouwman) [inline packages/common/src/search/searchCatalogs.ts:1]: This is the main fn for "Browse All" and "Omni-Search"

**Review by mjuniper — APPROVED** [review]

**Comment by dbouwman** [comment]
Changes:
- if an `IQuery` does not have type predicates, we should add all types for create and existing

**Comment by dbouwman** [comment]
Could not resolve outdated conversation - which was just me providing guidance on what to focus on...

**Pull request merged**